fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java

```diff
@@ -255,4 +402,57 @@ public interface PlayChannelHandler {
 		 */
 		void receive(MinecraftClient client, ClientPlayNetworkHandler handler, PacketByteBuf buf, PacketSender responseSender);
 	}
+
+	/**
+	 * An internal packet handler that works as a proxy between old and new API.
+	 * @param <T> the type of the packet
+	 */
+	private interface PlayChannelHandlerProxy<T extends FabricPacket> extends PlayChannelHandler {
+		PlayPacketHandler<T> getOriginalHandler();
+	}
+
+	/**
+	 * A thread-safe packet handler utilizing {@link FabricPacket}.
+	 * @param <T> the type of the packet
+	 */
+	@FunctionalInterface
+	public interface PlayPacketHandler<T extends FabricPacket> {
+		/**
+		 * Handles the incoming packet. This is called on the render thread, and can safely
+		 * call rendering-related methods.
+		 *
+		 * <p>An example usage of this is to display an overlay message:
+		 * <pre>{@code
+		 * // See FabricPacket for creating the packet
+		 * ClientPlayNetworking.registerReceiver(OVERLAY_PACKET_TYPE, (player, packet, responseSender) -> {
+		 * 	MinecraftClient.getInstance().inGameHud.setOverlayMessage(packet.message(), true);
+		 * });
+		 * }</pre>
+		 *
+		 * <p>The network handler can be accessed via {@link ClientPlayerEntity#networkHandler}.
+		 *
+		 * @param client the client
+		 * @param player the player that received the packet
+		 * @param packet the packet
+		 * @param responseSender the packet sender
+		 * @see FabricPacket
+		 */
+		void receive(MinecraftClient client, ClientPlayerEntity player, T packet, PacketSender responseSender);
+
+		/**
+		 * Processes an incoming packet. This is called on the network thread.
+		 * If this method returns {@code false}, {@link #receive} is not called.
+		 * This is for advanced users only; for normal uses, {@link #receive} should be
+		 * implemented instead.
+		 *
+		 * @param client the client
+		 * @param player the player that received the packet
+		 * @param packet the packet
+		 * @param responseSender the packet sender
+		 * @return whether to proceed to {@link #receive}
+		 */
+		default boolean processOnNetworkThread(MinecraftClient client, ClientPlayerEntity player, T packet, PacketSender responseSender) {
```

Comment:
Question: Do we want to pass some sort of context object instad of the client and player? Would make it easier to expand in the future without breaking?

Reply:
Also, the player is available from the client instance

Reply:
I believe our convention is not to pass context object unless it makes more sense (see DynamicRegistrySetupCallback).

The reason this passes player is simple - it passed `player` only first, realized `player.client` is protected, so I passed client as well; `client.player` does exist, but it is marked `@Nullable` (when, here, it is never null). I believe this is fine.

Reply:
Client is available from a static reference at `MinecraftClient.getInstance()`. I don't see any need for it to be passed at all.

Reply:
Contributing guideline specifies, "Consider passing the MinecraftClient instance as a parameter if it makes sense."

Reply:
Operative words being "if it makes sense". I explicitly ignore it in all of my networking code because it's just another client-exclusive class that I'd have to perform workarounds to reference in a way that won't break on the server.

Passing player makes sense, because I assume it's possible to receive packets for an instance of ClientOtherPlayerEntity as well (I hope), whereas passing client is never more than a convenience since it's always going to be accessible via the public non-nullavle, statics singleton reference (and is never not that).